src/com/twitter/intellij/pants/service/project/resolver/PantsSourceRootsExtension.java

```diff
@@ -67,6 +77,72 @@ public void resolve(
 
       createContentRoots(moduleDataNode, targetInfo);
     }
+    installSyntheticModules(executor, projectDataNode, modules);
+  }
+
+  /**
+   * Takes a set of paths. Returns a new set without paths that are children of any other path in the set
+   * Ex. for [A, A/B, A/B/C, B, C/D] the result set is [A, B, C/D]
+   */
+  public static Set<String> removeChildren(Set<String> paths) {
+    if(paths.size() > 0 ) {
+      List<String> sorted = paths.stream().sorted().collect(Collectors.toList());
+      String current = sorted.get(0);
+      List<String> buf = new ArrayList<>();
+      buf.add(current);
+      for (String s : sorted.stream().skip(1).collect(Collectors.toList())) {
+        if (!s.startsWith(current)) {
+          current = s;
+          buf.add(s);
+        }
+      }
+      return new HashSet<>(buf);
+    }
+    return Collections.emptySet();
+  }
+
+  private void installSyntheticModules(
+    @NotNull PantsCompileOptionsExecutor executor,
+    @NotNull DataNode<ProjectData> projectDataNode,
+    @NotNull Map<String, DataNode<ModuleData>> modules
+  ) {
+    Path projectPath = Paths.get(executor.getBuildRoot().getAbsolutePath());
+    Set<String> targetSpecPaths =
+      executor.getOptions().getSelectedTargetSpecs().stream()
+        .map(x -> x.replaceFirst("/?:.*", ""))
+        .map(x -> Paths.get(x).isAbsolute() ? projectPath.relativize(Paths.get(x)).toString() : x)
+        .collect(Collectors.toSet());
+    Set<String> targetSpecPathsWithoutChildren = removeChildren(targetSpecPaths);
+    Set<String> allContentRoots =
+      modules.values().stream()
+        .flatMap(x -> x.getChildren()
+        .stream())
+        .map(x -> x.getData(ProjectKeys.CONTENT_ROOT))
+        .filter(Objects::nonNull)
+        .map(x -> x.getRootPath())
+        .collect(Collectors.toSet());
+
+    // todo - optimize from m*n to m * log(n)
+    targetSpecPathsWithoutChildren.removeIf(targetSpec -> allContentRoots.stream().anyMatch(contentRoot ->  projectPath.resolve(targetSpec).startsWith(contentRoot)));
```

Comment:
Left a "todo" here because I'm not sure if it's worth doing it. I iterate over the list of all already created content roots for each target spec passed as an argument. This means N*M complexity. It could be optimized to N log(M) if I used trie data structure.